src/sgb-xi/validation.ts

```diff
@@ -2,37 +2,335 @@
+                      "zuschlag",
+                      "zuschlagszuordnung",
+                      "zuschlagsberechnung",
+                      "istAbzugStattZuschlag",
+                      "wert"
+                    ], ["invoices", invoiceIndex, "faelle", fallIndex, "einsaetze", einsatzIndex, "leistungen", leistungIndex, "zuschlaege", zuschlagIndex])
+                  )
+                )
+              ))
+            )
+          )
+        )
+      )
+    )
+  ).map(addMessage);
+
+  warnings = warnings.concat(
+    !billing?.abrechnungsstelle
+    ? []
+    : isTruncated(billing?.abrechnungsstelle, [
+        ["name", 30]
+      ], ["billing", "abrechnungsstelle"])
+      .concat(
+        isFailing([
+          billing?.abrechnungsstelle?.ansprechpartner?.every(ansprechpartner =>
+            Object.values(ansprechpartner).filter(Boolean).join(", ").length <= 30
+          ) || textIsTruncated("ansprechpartner", 30),
+        ], ["billing", "abrechnungsstelle"])
+      )
+  ).concat(
+    invoices?.flatMap((invoice, invoiceIndex) =>
+      isTruncated(invoice?.leistungserbringer, [
+        ["name", 30]
+      ], ["invoices", invoiceIndex, "leistungserbringer"])
+      .concat(
+        isFailing([
+          invoice?.leistungserbringer?.ansprechpartner?.every(ansprechpartner =>
+            Object.values(ansprechpartner).filter(Boolean).join(", ").length <= 30
+          ) || textIsTruncated("ansprechpartner", 30),
+        ], ["invoices", invoiceIndex, "leistungserbringer"])
+      )
+      .concat(
+        invoice?.faelle?.flatMap((fall, fallIndex) =>
+          isTruncated(fall?.versicherter, [
+            ["firstName", 45],
+            ["lastName", 45],
+            ["street", 46],
+            ["houseNumber", 9],
+            ["postalCode", 10],
+            ["city", 40],
+          ], ["invoices", invoiceIndex, "faelle", fallIndex, "versicherter"])
+          .concat(
+            fall?.einsaetze?.flatMap((einsatz, einsatzIndex) =>
+              einsatz?.leistungen?.flatMap((leistung, leistungIndex) =>
+                (!leistung?.hilfsmittel
+                  ? []
+                  : isTruncated(leistung?.hilfsmittel, [
+                      ["genehmigungskennzeichen", 15],
+                      ["bezeichnungPflegehilfsmittel", 30],
+                      ["inventarnummerPflegehilfsmittel", 15]
+                    ], ["invoices", invoiceIndex, "faelle", fallIndex, "einsaetze", einsatzIndex, "leistungen", leistungIndex, "hilfsmittel"])
+                ).concat(
+                  leistung?.zuschlaege?.flatMap((zuschlag, zuschlagIndex) => 
+                    isTruncated(zuschlag, [
+                      ["beschreibungZuschlagsart", 50]
+                    ], ["invoices", invoiceIndex, "faelle", fallIndex, "einsaetze", einsatzIndex, "leistungen", leistungIndex, "zuschlaege", zuschlagIndex])
+                  )
+                )
+              )
+            )
+          )
+        )
+      )
+    )
```

Comment:
Why are there warnings? I.e., why are not all warnings errors?

I figure you want to get all warnings and errors (instead of throwing an error on the first one) because it should all be shown simultaneously in the UI? Or is this maybe independent of UI, f.e. I would expect that the input field for fields with a max length already don't allow input longer than X.

I didn't more closely inspect the code because I have some issues with understanding code written in this style.

Reply:
Errors are show-stoppers. The user should never send the data in case of errors. These warnings exist to point out text that is truncated, but when it's fine to do so, so the user is aware of the shortcomings.

Yes, the errors and warnings shall be shown in the UI simultaneously at their origins. At the same time, I want the validation to catch errors for other consumers of the library. So max length validation will happen in two places for CoopCare, but I want to play it safe.

I agree again, this code is difficult to read. When I wrote it I was too much into "don't repeat yourself" instead of readability. I still want it to be concise and declarative, and I have some ideas to improve the readability partially, but I had to move on to get this done.

Reply:
👍 thanks

IMO "don't repeat yourself" + "readability" can both be achieved if ... well that sounds trivial but it is often not.... common/generic code is put into functions and the application logic code then uses these functions.

The kotlin standard library with functions like `associateBy` and many similar ones is IMO a good example of this pattern. In general, when I code, I try to reduce the number of nested scopes as much as possible, so the code looks "flat". F.e. Javascript's `async/await` but even before, the `.then.then.then` promises are an escape from the kind of async programming that ends up as a christmas tree.